feat: mamba support for mulled-build

johanneskoester opened the pull request:
This PR adds two command line args to mulled-build: `--use-mamba` and `--mamba-version`. While the latter is analogous to the already existing `--conda-version` argument, the former enables the use of mamba for package installation in mulled-build, instead of conda.

### Rationale

By also using Mamba in mulled, as it already happens in normal builds for Bioconda and conda-forge, we avoid discrepancies in the solver which make it very hard to debug sudden failures in the mulled build. Further, mamba is much faster in solving. Hopefully this will also help with a timeout that we currently see when trying to build bioconda-utils (https://github.com/bioconda/bioconda-recipes/pull/37331).

Above error makes this PR a bit pressing. A quick review, merge, and release would be highly appreciated.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

johanneskoester committed:
feat: mamba support for mulled-build

johanneskoester committed:
extend test case

mbargull commented on lib/galaxy/tool_util/deps/mulled/mulled_build.py:
```diff
@@ -284,9 +286,18 @@ def mull_targets(
+    else:
+        if conda_version is not None:
+            involucro_args.extend(["-set", f"PREINSTALL=conda install {verbose} --yes conda={conda_version}"])
+    involucro_args.extend(["-set", "CONDA_BIN=%s" % conda_bin])
```
```suggestion
    if use_mamba:
        conda_bin = "mamba"
        if mamba_version is None:
            mamba_version = ""
    involucro_args.extend(["-set", "CONDA_BIN=%s" % conda_bin])
    if conda_version is not None or mamba_version is not None:
        mamba_test = "true"
        specs = []
        if conda_version is not None:
            specs.append(f"conda={conda_version}")
        if mamba_version is not None:
            specs.append(f"mamba={mamba_version}")
            if mamba_version == "" and not specs:
                # If nothing but mamba without a specific version is requested,
                # then only run conda install if mamba is not already installed.
                mamba_test = f"[ '[]' = \"$( conda list --json --full-name mamba )\" ]"
        conda_install = f"""conda install {verbose} --yes {" ".join(f"'{spec}'" for spec in specs)}"""
        involucro_args.extend(["-set", f"PREINSTALL=if {mamba_test} ; then {conda_install} ; fi"]}
```
Changes:
- Decouple `use_mamba` and the version requirements (e.g., one could use `mamba`, but still require a specific `conda` version). Also, if both `conda_version` and `mamba_version` are given, still have only one `conda install` invocation.
- Even if `mamba_version` is `None`, we can set append an empty constraint `=` since `conda install` handles `mamba=` and `mamba` equivalently.
- Skip `conda install mamba` if `mamba` is already installed. This is, e.g., the case in the `quay.io/bioconda/create-env` container image we use in `bioconda-utils`.

mbargull commented on test/unit/tool_util/mulled/test_mulled_build.py:
```diff
@@ -27,11 +27,18 @@ def test_base_image_for_targets(target, version, base_image):
+@pytest.mark.parametrize("use_mamba", [False, True])
 @external_dependency_management
-def test_mulled_build_files_cli(tmpdir):
+def test_mulled_build_files_cli(target, tmpdir):
```
```suggestion
def test_mulled_build_files_cli(use_mamba, tmpdir):
```

mbargull reviewed (changes requested):
- `use_mamba` and the version requirements should be decoupled.
- Minor fix in the test.

mvdbeek committed:
Tweak use_mamba logic, skip install in mamba is installed

Decouple use_mamba and the version requirements (e.g., one could use mamba, but still require a specific conda version). Also, if both conda_version and mamba_version are given, still have only one conda install invocation.
Even if mamba_version is None, we can set append an empty constraint = since conda install handles mamba= and mamba equivalently.
Skip conda install mamba if mamba is already installed. This is, e.g., the case in the quay.io/bioconda/create-env container image we use in bioconda-utils.

Co-authored-by: Marcel Bargull <mbargull@users.noreply.github.com>

mvdbeek commented on lib/galaxy/tool_util/deps/mulled/mulled_build.py:
```diff
@@ -284,9 +286,28 @@ def mull_targets(
+            if mamba_version == "" and not specs:
+                # If nothing but mamba without a specific version is requested,
+                # then only run conda install if mamba is not already installed.
+                mamba_test = f"[ '[]' = \"$( conda list --json --full-name mamba )\" ]"
```
```suggestion
                mamba_test = "[ '[]' = \"$( conda list --json --full-name mamba )\" ]"
```

mvdbeek committed:
Fix parametrized test

Co-authored-by: Marcel Bargull <mbargull@users.noreply.github.com>

mvdbeek committed:
Drop unused f-string marker

mvdbeek committed:
Fix syntax error

mvdbeek committed:
Add conda-forge channel when installing mamba

mvdbeek reviewed (approved):
Looks good to me, thanks for the contribution @mbargull @johanneskoester.
@nsoranzo @bgruening should we merge ? Given this is all optional I don't see any harm.

mbargull commented on lib/galaxy/tool_util/deps/mulled/invfile.lua:
```diff
@@ -45,6 +45,7 @@ if conda_image == '' then
     conda_image = 'continuumio/miniconda3:latest'
```
```suggestion
    conda_image = 'quay.io/condaforge/mambaforge:latest'
```

mbargull replied:
When we use `continuumio/miniconda3:latest` and then do `conda install ... -c conda-forge`, then we re-install every package since we switch from `defaults` to `conda-forge`.

mbargull replied:
With `condaforge/mambaforge` we avoid that re-installation and also have `mamba` installed from the get go.
Caveat: `condaforge/mambaforge` explicitly has only `conda-forge` in the channel list, i.e., `default` removed. Personally, I think that's fine/good. However, if that change messes something up on your/someone's end, then we'd have to *always* run `conda config --append channels defaults` in the `PREINSTALL`.

nsoranzo replied:
Thanks, do we need this to be also a parameter for `mull_targets()` below? Currently this seems only configurable via the `CONDA_IMAGE` environment variable, unlike `DEST_BASE_IMAGE` which can be configured both as environment variable and with the `base_image` parameter.

mvdbeek replied:
Given https://github.com/galaxyproject/galaxy/pull/14770#issuecomment-1279969406 we could just make the switch. If anyone needs the older install image they can use a previous version of galaxy-tool-util, or let us know and we make it configurable.

mbargull replied:
I agree. I only brought up `defaults` removal so you're aware of it and because I don't know in which other contexts `mulled` is used around Galaxy. Generally, I'd endorse/recommend only using `conda-forge`+`bioconda` by now.

mvdbeek committed:
Switch conda install image, drop extra -c conda-forge
